plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts

```diff
@@ -99,106 +100,98 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
         }),
       );
     }
-    this.running = true;
-    this.run();
-  }
 
-  private async run() {
-    while (this.running) {
-      try {
-        // TODO: We want to disconnect the queue popping and message processing
-        // so that if the queue popping fails we exponentially back off in order to give the DB room to sort itself out.
-        await this.process();
-      } catch (e) {
-        this.logger.warn('Processing failed with:', e);
-        // TODO: this can be a little smarter as mentioned in the above comment.
-        // But for now, if something fails, wait a brief time to pick up the next message.
-        await this.wait();
-      }
+    if (this.stopFunc) {
+      throw new Error('Processing engine is already started');
     }
-  }
 
-  private async process() {
-    const { items } = await this.processingDatabase.transaction(async tx => {
-      return this.processingDatabase.getProcessableEntities(tx, {
-        processBatchSize: 1,
-      });
-    });
+    this.stopFunc = startTaskPipeline<RefreshStateItem>({
+      lowWatermark: 5,
+      highWatermark: 10,
+      loadTasks: async count => {
+        try {
+          const { items } = await this.processingDatabase.transaction(
+            async tx => {
+              return this.processingDatabase.getProcessableEntities(tx, {
```

Comment:
This getter really should not need a transaction. I see it is made out of two queries now but let's fix that (outside of this PR).

Reply:
Kinda nice to stick to the transaction pattern though tbh, otherwise you end up having to make the decision of whether a method should receive a transaction or not for each method.

Reply:
I just think that that decision (including isolation level and rollback strategy or whatnot) maybe belongs inside that method then - unless we explicitly want to compose it with other things in the same transaction. In that case maybe it should take an _optional_ transaction and use it if given, and otherwise do what it feels makes sense locally. Maybe that's a design we could be looking at overall later

Reply:
There's of course also the matter of raw speed - not sure how important it is here but there may be some gains to be had just like in the stitcher if we skip the transaction dance.